### PR TITLE
[benchmark] Add HF LLM benchmarks

### PR DESCRIPTION
**Pull request by angelayi**
Results in https://docs.google.com/spreadsheets/d/1xXOPg9JjEmPx0zc5QBNdyXQq8-K2_r4ybHaiS-q7pZ0/edit?gid=88695043#gid=88695043

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @Lucaskabela

### CODE REVIEW

File: torch/_dynamo/utils.py [hunk 1]
```diff
@@ -2118,17 +2118,25 @@ def clone_inputs(example_inputs):
...
         for key, value in res.items():
-            if isinstance(value, tuple):
+            if isinstance(value, (tuple, list, dict)):
                 res[key] = clone_inputs(value)
-            else:
-                assert isinstance(value, torch.Tensor), type(value)
+            elif isinstance(value, torch.Tensor):
                 res[key] = clone_input(value)
+            else:
+                res[key] = value
```
Comment (BoyuanFeng) [inline torch/_dynamo/utils.py]: Why do we loose the restriction here? Might be better to be explicit instead of implicit.

File: torch/_dynamo/utils.py [hunk 2]
```diff
@@ -2118,17 +2118,25 @@ def clone_inputs(example_inputs):
...
         return res
 
-    res = list(example_inputs)
-    for i in range(len(res)):
-        if isinstance(res[i], torch.Tensor):
-            res[i] = clone_input(res[i])
+    if isinstance(example_inputs, list):
+        res = list(example_inputs)
```
Comment (BoyuanFeng) [inline torch/_dynamo/utils.py]: seems duplicate

File: benchmarks/dynamo/huggingface_llm_models.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+        return model, dict(inputs)
+
+
+class TextGenerationBenchmark(Benchmark):
+    @staticmethod
+    def get_model_and_inputs(model_name, device):
+        tokenizer = AutoTokenizer.from_pretrained(model_name)
+        model = AutoModelForCausalLM.from_pretrained(
+            model_name,
```
Comment (anijain2305) [inline benchmarks/dynamo/huggingface_llm_models.py]: I think we have to use static_cache = True to enable compile.

File: benchmarks/dynamo/huggingface_llm.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+        
+        return device, model_name, model, example_inputs, batch_size
+
+    def iter_model_names(self, args):
+        yield from models.keys()
+    
+    def forward_pass(self, mod, inputs, collect_outputs=True):
+        with self.autocast(**self.autocast_arg):
+            torch.compiler.cudagraph_mark_step_begin()
+            res = mod.generate(**inputs)# , generation_config=self.generation_config)
```
Comment (BoyuanFeng) [inline benchmarks/dynamo/huggingface_llm.py]: nit

File: benchmarks/dynamo/huggingface_llm_models.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
+
```
Comment (BoyuanFeng) [inline benchmarks/dynamo/huggingface_llm_models.py]: nit

File: benchmarks/dynamo/huggingface_llm_models.py [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+class TextGenerationBenchmark(Benchmark):
+    @staticmethod
+    def get_model_and_inputs(model_name, device):
+        tokenizer = AutoTokenizer.from_pretrained(model_name)
+        model = AutoModelForCausalLM.from_pretrained(
+            model_name,
+        ).to(device)
+
+        prompt = "Once upon a time,"
```
Comment (BoyuanFeng) [inline benchmarks/dynamo/huggingface_llm_models.py]: should we use a more realistic input/output length for benchmarking? E.g., [vLLM](https://x.com/vllm_project/status/1913004324613243166) uses `200 in 200 out`, `1k in 2k out`, `5k in 1k out`, `10k in 500 out`, `30k in 100 out`. Currently we are using `4 in 50 out`, which may not reflect real-world workload of LLM inference.

File: benchmarks/dynamo/huggingface_llm.yaml [hunk 7]
```diff
@@ -0,0 +1,7 @@
+# removing this file prevents the Runner from dynamically skipping models
+only_inference:
+  - meta-llama/Llama-3.2-1B
```
Comment (BoyuanFeng) [inline benchmarks/dynamo/huggingface_llm.yaml]: could we merge huggingface_llm with other benchmarks (e.g., huggingface or blueberry)? So we can visualize on the dashboard without adding an extra column.

File: benchmarks/dynamo/huggingface_llm.py [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+        device,
+        model_name,
+        batch_size=None,
+        extra_args=None,
+    ):
+
+        benchmark_cls = models[model_name]
+        model, example_inputs = benchmark_cls.get_model_and_inputs(model_name, device)
+
+        batch_size = 1
```
Comment (BoyuanFeng) [inline benchmarks/dynamo/huggingface_llm.py]: should we run with larger batch size (e.g., 8)? although bs=1 is also meaningful for users who run local inference.

**Comment by BoyuanFeng** [comment]
Thanks for adding more models! A few minor comments. Also, please fix the ci.

File: benchmarks/dynamo/huggingface_llm_models.py (hunk 3 above)
Comment (StrongerXi) [reply benchmarks/dynamo/huggingface_llm_models.py]: Do we compile with `fullgraph=True`?

Also I found out about [this](https://github.com/huggingface/transformers/blob/8c4ea670dceace8d9b1bac8310bc62146b7134cd/src/transformers/models/qwen3_moe/modeling_qwen3_moe.py#L409), not sure if it's relevant here.

But on a tangential note, maybe we should start thinking about how to trace/optimize MoEs (e.g., some inductor pass to pattern match & fuse grouped gemms, and maybe prologue fusion for the activation gather). Right now a lot of the impls out there aren't very trace-friendly due to [routing dynamisms](https://github.com/huggingface/transformers/blob/8c4ea670dceace8d9b1bac8310bc62146b7134cd/src/transformers/models/qwen3_moe/modeling_qwen3_moe.py#L247-L256).

MoE is pretty standard now in LLMs, and is making its way into image generation ([HiDream](https://github.com/HiDream-ai/HiDream-I1/blob/5f92bab45f1dfb1e794ee357286a5b837eaf4400/hi_diffusers/models/moe.py#L136-L154)) and video generation (Wan2.2, although a very [unconventional](https://github.com/Wan-Video/Wan2.2/blob/031a9be56cec91e86d140d3d3a74280fb05a9b1c/wan/text2video.py#L186-L201) MoE). Gpt-fast does mention it, but the implementation is pretty limited and not practical imo (only works for small batch and seqlen) [1](https://github.com/pytorch-labs/gpt-fast/blob/7dd5661e2adf2edd6a1042a2732dcd3a94064ad8/mixtral-moe/model.py#L212-L219), [2](https://www.thonking.ai/p/short-supporting-mixtral-in-gpt-fast/comment/50424941).

Comment (BoyuanFeng) [reply benchmarks/dynamo/huggingface_llm_models.py]: for moe part, could you file an issue? we can separate discussions on benchmark & optimization

**Comment by BoyuanFeng** [comment]
Curious, will we add these models into existing Huggingface column or a new column called "huggingface_llm"? It might be a bit confusing with two columns starting with "huggingface"..


image

**Comment by angelayi** [comment]
@BoyuanFeng yes! I have updated to merge everything into the huggingface column.

File: benchmarks/dynamo/common.py [hunk 9]
```diff
@@ -1098,14 +1098,27 @@ def maybe_mark_profile(*args, **kwargs):
...
 
     with maybe_profile(args.export_profiler_trace, **args.profile_details) as p:
-        if args.export_aot_inductor:
-            frozen_model_iter_fn = export_aot_inductor(
-                model, example_inputs, args.inductor_compile_mode
-            )
-        elif args.export_nativert:
-            frozen_model_iter_fn = export_nativert(model, example_inputs)
+        use_generate_mode = kwargs.get("use_generate_mode", False)
+        if use_generate_mode:
```
Comment (angelayi) [inline benchmarks/dynamo/common.py]: I added this use_generate_mode flag so that we only apply torch.compile/export to model.forward, instead of applying it to model.generate

File: benchmarks/dynamo/huggingface_llm_models.py [hunk 10]
```diff
@@ -0,0 +1,96 @@
+import subprocess
```
Comment (huydhn) [inline benchmarks/dynamo/huggingface_llm_models.py:1]: For the logic in this new classes, I think we should eventually have them in https://github.com/pytorch/benchmark/blob/main/torchbenchmark/util/framework/huggingface/model_factory.py and bring it to CI by bumping TorchBench pinned commit.  To be fair, updating TorchBench on PyTorch CI hasn't been done for a while, and I'm reviving this process (https://github.com/pytorch/pytorch/pull/161461).  So, instead of waiting for that, this is probably ok to keep this file here for now to unblock the new models.  Then, we can add it to TorchBench.  Any thoughts?

Comment (angelayi) [reply benchmarks/dynamo/huggingface_llm_models.py:1]: sounds good!

File: benchmarks/dynamo/ci_expected_accuracy/inductor_huggingface_inference.csv [hunk 11]
```diff
@@ -171,3 +171,18 @@ XLNetLMHeadModel,pass,0
 
 
 YituTechConvBert,pass,0
+
+
+meta-llama/Llama-3.2-1B,fail_accuracy,0
```
Comment (huydhn) [inline benchmarks/dynamo/ci_expected_accuracy/inductor_huggingface_inference.csv]: Btw, for the lint failure annoyance, you need 3 line gap here between each entry.  Just FYI, this is to trick git conflict resolution

File: benchmarks/dynamo/huggingface.py [hunk 12]
```diff
@@ -440,14 +443,39 @@ def load_model(
...
+                model_name, device
+            )
 
-        # So we can check for correct gradients without eliminating the dropout computation
-        for attr in dir(config):
-            if "drop" in attr and isinstance(getattr(config, attr), float):
-                setattr(config, attr, 1e-30)
+            # Set this flag so that when we test for speedup, we use
+            # model.generate instead of using model.forward
+            self.hf_llm = True
```
Comment (huydhn) [inline benchmarks/dynamo/huggingface.py:456]: Default `hf_llm` to false to avoid the failure in https://github.com/pytorch/pytorch/actions/runs/17278673679/job/49047949664?

**Comment by huydhn** [comment]
@pytorchbot rebase -b main

(to get the latest TorchBench bump from https://github.com/pytorch/pytorch/pull/161461)

**Comment by huydhn** [comment]
I'm running a round of HF benchmark at https://github.com/pytorch/pytorch/actions/runs/17597039052, let's see how it goes

**Review by huydhn — APPROVED** [review]
LGTM!  Thank you for adding these models.  Let's finish a round of benchmark https://github.com/pytorch/pytorch/actions/runs/17597039052 to confirm

File: benchmarks/dynamo/huggingface.py [hunk 13]
```diff
@@ -530,7 +560,9 @@ def compute_loss(self, pred):
 
     def forward_pass(self, mod, inputs, collect_outputs=True):
         with self.autocast(**self.autocast_arg):
-            return mod(**inputs)
+            res = mod(**inputs)
+        if self.hf_llm:
```
Comment (huydhn) [inline benchmarks/dynamo/huggingface.py]: Found a small bug here, this should be `return res.logits if self.hf_llm else res`

**Comment by huydhn** [comment]
IMO, we can ignore those ROCm failures and update them in a subsequence PR.  Some ROCm failures were from trunk and would need another rebase to get rid of.

File: benchmarks/dynamo/huggingface.py (hunk 13 above)
Comment (huydhn) [reply benchmarks/dynamo/huggingface.py]:
```suggestion
        return res.logits if self.hf_llm else res
```

File: benchmarks/dynamo/huggingface.yaml [hunk 14]
```diff
@@ -9,9 +9,16 @@ skip:
     # Fails with even batch size = 1
     - GPTJForCausalLM
     - GPTJForQuestionAnswering
+    # Model too big
+    - google/gemma-3-4b-it
```
Comment (huydhn) [inline benchmarks/dynamo/huggingface.yaml:13]: Once this lands, let me see if we could use a100 for all HF models instead.  This should resolve this issue

File: benchmarks/dynamo/huggingface.py [hunk 15]
```diff
@@ -440,14 +444,46 @@ def load_model(
...
+
+            # So we can check for correct gradients without eliminating the dropout computation
+            for attr in dir(config):
+                if "drop" in attr and isinstance(getattr(config, attr), float):
+                    setattr(config, attr, 1e-30)
+
+            # Turning off kv cache for torchbench models. This is not the right
+            # thing to do, but the pt2 dashboard is outdated. Real transformers
+            # benchmarks will be added soon using a different infra.
+            if hasattr(model, "config") and hasattr(model.config, "use_cache"):
```
Comment (huydhn) [inline benchmarks/dynamo/huggingface.py:482]: I'm see this error on `aot_inductor` job https://github.com/pytorch/pytorch/actions/runs/17666799214/job/50233769713#step:24:1966.  If my understand is correct, we could disable cache on LLM models for now to fix it?

Comment (huydhn) [reply benchmarks/dynamo/huggingface.py:482]: Chat with Angela, will just update the expect value instead
